components/scheduler/scheduler.go

```diff
@@ -0,0 +1,414 @@
+		continuedFailure = nil
+	}
+
+	// TODO(Tianyu): Add callback here
+	req := &workflowservice.StartWorkflowExecutionRequest{
+		WorkflowId:               workflowID,
+		WorkflowType:             newWorkflow.WorkflowType,
+		TaskQueue:                newWorkflow.TaskQueue,
+		Input:                    newWorkflow.Input,
+		WorkflowExecutionTimeout: newWorkflow.WorkflowExecutionTimeout,
+		WorkflowRunTimeout:       newWorkflow.WorkflowRunTimeout,
+		WorkflowTaskTimeout:      newWorkflow.WorkflowTaskTimeout,
+		Identity:                 s.identity(),
+		RequestId:                uuid.NewString(),
+		WorkflowIdReusePolicy:    enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE,
+		RetryPolicy:              newWorkflow.RetryPolicy,
+		Memo:                     newWorkflow.Memo,
+		// TODO(Tianyu): Search is not implemented for now
+		SearchAttributes:     &commonpb.SearchAttributes{IndexedFields: nil},
+		Header:               newWorkflow.Header,
+		LastCompletionResult: lastCompletionResult,
+		ContinuedFailure:     continuedFailure,
+		Namespace:            s.Args.State.Namespace,
+	}
+
+	res, err := s.frontendClient.StartWorkflowExecution(context.Background(), req)
+	if err != nil {
+		// TODO(Tianyu): original implementation translates this error which may not be relevant any more
+		return nil, err
+	}
+	// this will not match the time in the workflow execution started event
+	// exactly, but it's just informational so it's close enough.
+	now := time.Now()
+
+	if !start.Manual {
+		// record metric only for _scheduled_ actions, not trigger/backfill, otherwise it's not meaningful
+		desiredTime := cmp.Or(start.DesiredTime, start.ActualTime)
+		s.metrics.Timer(metrics.ScheduleActionDelay.Name()).Record(now.Sub(desiredTime.AsTime()))
+	}
+
+	return &schedpb.ScheduleActionResult{
+		ScheduleTime: start.ActualTime,
+		ActualTime:   timestamppb.New(now),
+		StartWorkflowResult: &commonpb.WorkflowExecution{
+			WorkflowId: workflowID,
+			RunId:      res.RunId,
+		},
+	}, nil
+}
+
+func (s *Scheduler) terminateWorkflow(ex *commonpb.WorkflowExecution) {
+	// TODO: remove after https://github.com/temporalio/sdk-go/issues/1066
+	// TODO(Tianyu): hardcoded wait time
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	rreq := &historyservice.TerminateWorkflowExecutionRequest{
+		NamespaceId: s.Args.State.NamespaceId,
+		TerminateRequest: &workflowservice.TerminateWorkflowExecutionRequest{
+			Namespace: s.Args.State.Namespace,
+			// only set WorkflowId so we cancel the latest, but restricted by FirstExecutionRunId
+			WorkflowExecution:   &commonpb.WorkflowExecution{WorkflowId: ex.WorkflowId},
+			Reason:              "terminated by schedule overlap policy",
+			Identity:            s.identity(),
+			FirstExecutionRunId: ex.RunId,
+		},
+	}
+	_, err := s.historyClient.TerminateWorkflowExecution(ctx, rreq)
+
+	// TODO(Tianyu): original implementation translates this error which may not be relevant any more
```

Comment:
I need to understand the semantics here better but just logging and adding here seems wrong to me.
Will we automatically retry in this case?

Reply:
This is an interesting question (to me, maybe you've figured it out): in the workflow impl we can take advantage of activity retries and even local activity retries, using the various workflow/activity/local activity state machines. What do we do here? I assume that `historyservice` client is a retriable client, but that's just a few quick retries. What if we want to e.g. back off and retry for up to an hour?

Reply:
We would need to store the failure state and number of attempts and start a timer task for the next attempt.
We have a similar mechanism for nexus operations and callbacks.

Reply:
I think just the machinery for starting a workflow and waiting for the callback could be its own separate nested machine. The scheduler can spawn multiple of these if the overlap policy permits it.

Reply:
I am going to add a comment there documenting this but that seems like a larger HSM improvement and is probably out of the scope of this PR. Will spend some time on designing this along with HSM signaling mechanism.